### ISSUE TITLE
HS can end up trying to insert into event_to_state_groups with a null state group

### ISSUE THREAD

**Issue by ara4n** [matrix-org/synapse#2730]
From ash:

A mix of:

```
Dec 05 06:41:00 helios python[523]: 2017-12-05 06:41:00,957 - synapse.federation.federation_server - 199 - ERROR - PUT-283198- Failed to handle PDU $15124740111137549SVpls:matrix.org
Dec 05 06:41:00 helios python[523]: Traceback (most recent call last):
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/federation/federation_server.py", line 191, in process_pdus_for_room
Dec 05 06:41:00 helios python[523]:     transaction.origin, pdu
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
Dec 05 06:41:00 helios python[523]:     result = result.throwExceptionIntoGenerator(g)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
Dec 05 06:41:00 helios python[523]:     return g.throw(self.type, self.value, self.tb)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/federation/federation_server.py", line 584, in _handle_received_pdu
Dec 05 06:41:00 helios python[523]:     yield self.handler.on_receive_pdu(origin, pdu, get_missing=True)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
Dec 05 06:41:00 helios python[523]:     result = result.throwExceptionIntoGenerator(g)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
Dec 05 06:41:00 helios python[523]:     return g.throw(self.type, self.value, self.tb)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/handlers/federation.py", line 193, in on_receive_pdu
Dec 05 06:41:00 helios python[523]:     origin, pdu, prevs, min_depth
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
Dec 05 06:41:00 helios python[523]:     result = result.throwExceptionIntoGenerator(g)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
Dec 05 06:41:00 helios python[523]:     return g.throw(self.type, self.value, self.tb)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/handlers/federation.py", line 313, in _get_missing_events_for_pdu
Dec 05 06:41:00 helios python[523]:     get_missing=False
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
Dec 05 06:41:00 helios python[523]:     result = result.throwExceptionIntoGenerator(g)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
Dec 05 06:41:00 helios python[523]:     return g.throw(self.type, self.value, self.tb)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/handlers/federation.py", line 237, in on_receive_pdu
Dec 05 06:41:00 helios python[523]:     auth_chain=auth_chain,
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
Dec 05 06:41:00 helios python[523]:     result = result.throwExceptionIntoGenerator(g)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
Dec 05 06:41:00 helios python[523]:     return g.throw(self.type, self.value, self.tb)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/handlers/federation.py", line 394, in _process_received_pdu
Dec 05 06:41:00 helios python[523]:     state=state,
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
Dec 05 06:41:00 helios python[523]:     result = result.throwExceptionIntoGenerator(g)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
Dec 05 06:41:00 helios python[523]:     return g.throw(self.type, self.value, self.tb)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/handlers/federation.py", line 1444, in _handle_new_event
Dec 05 06:41:00 helios python[523]:     auth_events=auth_events,
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
Dec 05 06:41:00 helios python[523]:     result = result.throwExceptionIntoGenerator(g)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
Dec 05 06:41:00 helios python[523]:     return g.throw(self.type, self.value, self.tb)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/handlers/federation.py", line 1610, in _prep_event
Dec 05 06:41:00 helios python[523]:     event, context.prev_state_ids, for_verification=True,
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
Dec 05 06:41:00 helios python[523]:     result = result.throwExceptionIntoGenerator(g)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
Dec 05 06:41:00 helios python[523]:     return g.throw(self.type, self.value, self.tb)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/api/auth.py", line 567, in compute_auth_events
Dec 05 06:41:00 helios python[523]:     join_rule_event = yield self.store.get_event(join_rule_event_id)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1386, in _inlineCallbacks
Dec 05 06:41:00 helios python[523]:     result = g.send(result)
Dec 05 06:41:00 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/events.py", line 608, in get_event
Dec 05 06:41:00 helios python[523]:     raise SynapseError(404, "Could not find event %s" % (event_id,))
Dec 05 06:41:00 helios python[523]: SynapseError: 404: Could not find event $143524675924487ENJWh:matrix.org
```

and

```
Dec 05 06:41:07 helios python[523]: 2017-12-05 06:41:07,222 - synapse.handlers.federation - 231 - ERROR - PUT-283228- Failed to get state for event: $15124736261133468acSTH:matrix.org
Dec 05 06:41:07 helios python[523]: Traceback (most recent call last):
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/handlers/federation.py", line 228, in on_receive_pdu
Dec 05 06:41:07 helios python[523]:     origin, pdu.room_id, pdu.event_id,
Dec 05 06:41:07 helios python[523]: ResponseFailed: [<twisted.python.failure.Failure twisted.web._newclient.ConnectionAborted: >, <twisted.python.failure.Failure twisted.web.http._DataLoss: >]
Dec 05 06:41:07 helios python[523]: 2017-12-05 06:41:07,317 - synapse.storage.events - 189 - ERROR - PUT-283228- IntegrityError, retrying.
Dec 05 06:41:07 helios python[523]: Traceback (most recent call last):
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/events.py", line 187, in f
Dec 05 06:41:07 helios python[523]:     res = yield func(self, *args, **kwargs)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
Dec 05 06:41:07 helios python[523]:     result = result.throwExceptionIntoGenerator(g)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
Dec 05 06:41:07 helios python[523]:     return g.throw(self.type, self.value, self.tb)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/events.py", line 395, in _persist_events
Dec 05 06:41:07 helios python[523]:     new_forward_extremeties=new_forward_extremeties,
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
Dec 05 06:41:07 helios python[523]:     result = result.throwExceptionIntoGenerator(g)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
Dec 05 06:41:07 helios python[523]:     return g.throw(self.type, self.value, self.tb)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 325, in runInteraction
Dec 05 06:41:07 helios python[523]:     inner_func, *args, **kwargs
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 250, in inContext
Dec 05 06:41:07 helios python[523]:     result = inContext.theWork()
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
Dec 05 06:41:07 helios python[523]:     inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/context.py", line 122, in callWithContext
Dec 05 06:41:07 helios python[523]:     return self.currentContext().callWithContext(ctx, func, *args, **kw)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/context.py", line 85, in callWithContext
Dec 05 06:41:07 helios python[523]:     return func(*args,**kw)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 307, in _runWithConnection
Dec 05 06:41:07 helios python[523]:     compat.reraise(excValue, excTraceback)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 298, in _runWithConnection
Dec 05 06:41:07 helios python[523]:     result = func(conn, *args, **kw)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 319, in inner_func
Dec 05 06:41:07 helios python[523]:     func, *args, **kwargs
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 247, in _new_transaction
Dec 05 06:41:07 helios python[523]:     r = func(txn, *args, **kwargs)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/util/logutils.py", line 84, in wrapped
Dec 05 06:41:07 helios python[523]:     return f(*args, **kwargs)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/events.py", line 713, in _persist_events_txn
Dec 05 06:41:07 helios python[523]:     self._store_mult_state_groups_txn(txn, events_and_contexts)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/state.py", line 314, in _store_mult_state_groups_txn
Dec 05 06:41:07 helios python[523]:     for event_id, state_group_id in state_groups.iteritems()
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 476, in _simple_insert_many_txn
Dec 05 06:41:07 helios python[523]:     txn.executemany(sql, vals)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 87, in executemany
Dec 05 06:41:07 helios python[523]:     self._do_execute(self.txn.executemany, sql, *args)
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 114, in _do_execute
Dec 05 06:41:07 helios python[523]:     sql, *args
Dec 05 06:41:07 helios python[523]: IntegrityError: null value in column "state_group" violates not-null constraint
Dec 05 06:41:07 helios python[523]: DETAIL:  Failing row contains ($15124736261133468acSTH:matrix.org, null).
Dec 05 06:41:07 helios python[523]: 2017-12-05 06:41:07,435 - synapse.federation.federation_server - 199 - ERROR - PUT-283228- Failed to handle PDU $15124740521138053rtdzE:matrix.org
Dec 05 06:41:07 helios python[523]: Traceback (most recent call last):
Dec 05 06:41:07 helios python[523]:   File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/federation/federation_server.py", line 191, in process_pdus_for_room
Dec 05 06:41:07 helios python[523]:     transaction.origin, pdu
Dec 05 06:41:07 helios python[523]: IntegrityError: null value in column "state_group" violates not-null constraint
Dec 05 06:41:07 helios python[523]: DETAIL:  Failing row contains ($15124736261133468acSTH:matrix.org, null).
Dec 05 06:41:07 helios python[523]: 2017-12-05 06:41:07,437 - twisted - 131 - CRITICAL - - Unhandled error in Deferred:
Dec 05 06:41:07 helios python[523]: 2017-12-05 06:41:07,437 - twisted - 131 - CRITICAL - -
Dec 05 06:41:07 helios python[523]: IntegrityError: null value in column "state_group" violates not-null constraint
Dec 05 06:41:07 helios python[523]: DETAIL:  Failing row contains ($15124736261133468acSTH:matrix.org, null).
```

**Comment by turt2live** [comment]
Also getting this:

```
2018-03-27 08:20:18,933 - synapse.storage.events - 189 - ERROR - PUT-121- IntegrityError, retrying.
Traceback (most recent call last):
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/events.py", line 187, in f
    res = yield func(self, *args, **kwargs)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/events.py", line 395, in _persist_events
    new_forward_extremeties=new_forward_extremeties,
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 319, in runInteraction
    inner_func, *args, **kwargs
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 250, in inContext
    result = inContext.theWork()
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 307, in _runWithConnection
    compat.reraise(excValue, excTraceback)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 298, in _runWithConnection
    result = func(conn, *args, **kw)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 313, in inner_func
    func, *args, **kwargs
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 241, in _new_transaction
    r = func(txn, *args, **kwargs)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/util/logutils.py", line 84, in wrapped
    return f(*args, **kwargs)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/events.py", line 713, in _persist_events_txn
    self._store_mult_state_groups_txn(txn, events_and_contexts)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/state.py", line 715, in _store_mult_state_groups_txn
    for event_id, state_group_id in state_groups.iteritems()
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 470, in _simple_insert_many_txn
    txn.executemany(sql, vals)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 85, in executemany
    self._do_execute(self.txn.executemany, sql, *args)
  File "/home/synapse/.synapse/local/lib/python2.7/site-packages/synapse/storage/_base.py", line 112, in _do_execute
    sql, *args
IntegrityError: null value in column "state_group" violates not-null constraint
DETAIL:  Failing row contains ($Feldspar_Granite_Pie's_Lovely_Tiger_Assorts_Toby's_Strange_Mosq..., null).
```

**Comment by richvdh** [comment]
I think this probably got fixed by some work I did on `EventContext` (#6320 etc)
